### PR TITLE
Throw error if sending non-bencodable message

### PR DESCRIPTION
**Pull request by sanjayl**
Before submitting a PR make sure the following things have been done:

- [x] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [x] You've added tests to cover your change(s)
- [x] All tests are passing
- [x] The new code is not generating reflection warnings
- [ ] You've updated the readme (if adding/changing middleware)

Thanks!

Here's a first draft of something that will report back an error if the situation in #332 is detected. I tried fixing it directly in tools.nrepl, but I was having a tough time working through that code. This seems a bit hacky, but I think it should work. Let me know if you have more elegant ideas on working around this (ways to fix the issue in tools.nrepl, making our own transport, etc).

This tries to work around the hung communication related to sending
non-bencodable information through the standard Transport. If
non-bencodable information is currently sent in the reply message, it
can only be assigned to the value to the :value key, otherwise, the
Transport.send will silently fail and seize up. This patch will throw an
error to the user if this type of situation is detected.

### CODE REVIEW

**Comment by jgeraerts** [comment]
This would also handle #330 gracefully.

File: src/cider/nrepl/middleware/util/error_handling.clj [hunk 1]
```diff
@@ -70,6 +73,37 @@
         (coll? input)  :status-coll
         (keyword? input) :status-item))
 
+(defn- shallow-bencodable?
+  "Returns false if `item`'s type can't be bencoded as defined by the
+  algorithm in `clojure.tools.nrepl.bencode/write-bencode`. Does not
+  traverse the subelements of a collection to ensure that the
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/util/error_handling.clj]: subelements? Should this be just elements?

File: src/cider/nrepl/middleware/util/error_handling.clj [hunk 2]
```diff
@@ -97,6 +131,7 @@
 
 (defmethod op-handler :inline-reply [answer msg]
   (let [terminal-status (set/union #{:done} (normalize-status (:status answer)))]
+    (deep-bencodable-or-fail (dissoc answer :value))
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/util/error_handling.clj]: Probably you should add a comment explaining this.

File: src/cider/nrepl/middleware/util/error_handling.clj [hunk 3]
```diff
@@ -70,6 +73,37 @@
...
+    (string? item)  :string
+    (symbol? item)  :named
+    (keyword? item) :named
+    (map? item)     :map
+    (or (nil? item) (coll? item) (.isArray (class item))) :list
+    :else false))
+
+(defn- deep-bencodable-or-fail
+  "Walks through the object provided by `item` and returns true if
+  the entity and all subelements are bencodable as defined by the
```
Comment (bbatsov) [inline src/cider/nrepl/middleware/util/error_handling.clj]: subelements -> elements

**Comment by bbatsov** [comment]
Seems like a reasonable approach, although in the long term this should still be addressed in nREPL. @cemerick Any thoughts?

File: src/cider/nrepl/middleware/util/error_handling.clj (hunk 1 above)
Comment (sanjayl) [inline src/cider/nrepl/middleware/util/error_handling.clj]: Done

File: src/cider/nrepl/middleware/util/error_handling.clj (hunk 2 above)
Comment (sanjayl) [inline src/cider/nrepl/middleware/util/error_handling.clj]: Done

**Comment by bbatsov** [comment]
@sanjayl Thanks! Guess you can also file an upstream ticket, so we don't forget about an upstream solution to the problem.

**Comment by sanjayl** [comment]
Done! http://dev.clojure.org/jira/browse/NREPL-81

**Pull request merged**